src/wallet/wallet.cpp

```diff
@@ -2903,22 +2903,28 @@ bool CWallet::AttachChain(const std::shared_ptr<CWallet>& walletInstance, interf
 
     if (tip_height && *tip_height != rescan_height)
     {
-        if (chain.havePruned()) {
```

Comment:
in 9bf1245764579302fef8d86fbbe1003fc2bb7de4 "wallet: avoid rescans if under the snapshot"

I think we should retain this check and just expand it to also check `hasAssumedValidChain`. Otherwise we will run the while loop below every time a wallet is loaded, and that can have a significant effect on loading times.

With a testnet wallet where the `rescan_height` is set to 0, the wallet took 500 ms to load on master, while it took 5 secs with this PR.

Reply:
Good catch, thanks.